File: app/src/main/java/com/dougkeen/bart/activities/TripInProgressActivity.java
package com.dougkeen.bart.activities;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.text.format.DateFormat;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.core.content.IntentCompat;

import com.dougkeen.bart.BartRunnerApplication;
import com.dougkeen.bart.R;
import com.dougkeen.bart.data.LifecycleFlowCollector;
import com.dougkeen.bart.model.Departure;
import com.dougkeen.bart.model.Line;
import com.dougkeen.bart.model.TimeSource;
import com.dougkeen.bart.model.TripLeg;
import com.dougkeen.bart.model.TripStop;
import com.dougkeen.bart.platform.DepartureParcel;
import com.dougkeen.bart.presentation.DepartureTextFormatter;
import com.dougkeen.bart.services.BoardedDepartureService;

import java.util.Date;
import java.util.List;

/** Shows the live state of a selected, possibly multi-train trip. */
public class TripInProgressActivity extends AbstractViewActivity {

    private static final int POST_NOTIFICATIONS_REQUEST_CODE = 1002;

    private final Handler mHandler = new Handler(android.os.Looper.getMainLooper());
    private TimeSource mTimeSource;
    private Departure mDeparture;
    private TripProgressViewModel mTripProgressViewModel;
    private TripActionsViewModel mTripActionsViewModel;
    private boolean mIsFollowing;

    private TextView mStatus;
    private TextView mRoute;
    private TextView mArrival;
    private View mFollowTripButton;
    private LinearLayout mTimeline;

    private final Runnable mRefreshRunnable = new Runnable() {
        @Override
        public void run() {
            renderTrip();
            if (!isFinishing()) {
                mHandler.postDelayed(this, 1000L);
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.trip_in_progress);

        mStatus = (TextView) findViewById(R.id.tripStatus);
        mRoute = (TextView) findViewById(R.id.tripRoute);
        mArrival = (TextView) findViewById(R.id.tripArrival);
        mFollowTripButton = findViewById(R.id.followTripButton);
        mTimeline = (LinearLayout) findViewById(R.id.tripTimeline);

        mFollowTripButton.setOnClickListener(view -> followTrip());

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        BartRunnerApplication application = (BartRunnerApplication) getApplication();
        mTimeSource = application.getTimeSource();
        mTripActionsViewModel = new androidx.lifecycle.ViewModelProvider(this)
                .get(TripActionsViewModel.class);
        DepartureParcel departureParcel = IntentCompat.getParcelableExtra(
                getIntent(), BoardedDepartureService.DEPARTURE_EXTRA,
                DepartureParcel.class);
        Departure requestedDeparture = departureParcel == null
                ? null : departureParcel.getDeparture();
        if (requestedDeparture != null) {
            mDeparture = requestedDeparture;
            Departure followedDeparture = mTripActionsViewModel.getFollowedDeparture();
            mIsFollowing = followedDeparture != null
                    && followedDeparture.equals(requestedDeparture);
        } else {
            mDeparture = mTripActionsViewModel.getFollowedDeparture();
            mIsFollowing = mDeparture != null;
        }
        if (mDeparture == null) {
            finish();
            return;
        }
        mTripProgressViewModel = new androidx.lifecycle.ViewModelProvider(this)
                .get(TripProgressViewModel.class);
        mTripProgressViewModel.setDeparture(mDeparture, mTimeSource);
        LifecycleFlowCollector.collect(this,
                mTripProgressViewModel.getDepartureState(), departure -> {
                    if (departure == null) {
                        return;
                    }
                    mDeparture = departure;
                    if (mIsFollowing) {
                        mTripActionsViewModel.updateFollowedTrip(mDeparture);
                    }
                    renderTrip();
                    invalidateOptionsMenu();
                });
        updateFollowState();
        renderTrip();
    }

    @Override
    protected void onStart() {
        super.onStart();
        mHandler.post(mRefreshRunnable);
    }

    @Override
    protected void onStop() {
        mHandler.removeCallbacks(mRefreshRunnable);
        super.onStop();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.trip_in_progress_menu, menu);
        updateMenu(menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        updateMenu(menu);
        return super.onPrepareOptionsMenu(menu);
    }

    private void updateMenu(Menu menu) {
        if (mDeparture == null) {
            return;
        }
        MenuItem cancel = menu.findItem(R.id.cancel_alarm_button);
        MenuItem set = menu.findItem(R.id.set_alarm_button);
        MenuItem delete = menu.findItem(R.id.delete);
        boolean alarmPending = mTripActionsViewModel.isAlarmPending();
        cancel.setVisible(mIsFollowing && alarmPending);
        set.setVisible(mIsFollowing && !alarmPending
                && mDeparture.getMeanSecondsLeft(
                mDeparture.getMinEstimate(), mDeparture.getMaxEstimate(),
                mTimeSource.nowMillis()) > 60);
        delete.setVisible(mIsFollowing);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        if (mDeparture == null) {
            return super.onOptionsItemSelected(item);
        }
        if (item.getItemId() == R.id.set_alarm_button) {
            new com.dougkeen.bart.activities.TrainAlarmDialogFragment()
                    .show(getSupportFragmentManager(),
                            TrainAlarmDialogFragment.TAG);
            return true;
        } else if (item.getItemId() == R.id.cancel_alarm_button) {
            sendServiceAction(mTripActionsViewModel.cancelAlarm());
            return true;
        } else if (item.getItemId() == R.id.share_arrival) {
            Intent share = new Intent(Intent.ACTION_SEND);
            share.setType("text/plain");
            share.putExtra(Intent.EXTRA_SUBJECT, "My BART trip");
            share.putExtra(Intent.EXTRA_TEXT, getString(
                    R.string.arrival_message,
                    mDeparture.getStationPair().getDestination().getName(),
                    DepartureTextFormatter.estimatedArrivalTime(
                            this, mDeparture, false)));
            startActivity(Intent.createChooser(share,
                    getString(R.string.share_arrival_time)));
            return true;
        } else if (item.getItemId() == R.id.delete) {
            new AlertDialog.Builder(this)
                    .setMessage(R.string.clear_trip_confirmation)
                    .setNegativeButton(R.string.cancel, null)
                    .setPositiveButton(R.string.delete,
                            new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog,
                                                     int which) {
                                    sendServiceAction(mTripActionsViewModel.clearTrip());
                                    finish();
                                }
                            }).show();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void sendServiceAction(TripServiceCommand command) {
        Intent intent = new Intent(this, BoardedDepartureService.class)
                .setAction(command.getAction());
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            startForegroundService(intent);
        } else {
            startService(intent);
        }
    }

    private void followTrip() {
        if (mDeparture == null || mIsFollowing) {
            return;
        }

        TripServiceCommand command = mTripActionsViewModel.followTrip(mDeparture);
        requestNotificationPermissionIfNeeded();

        startBoardedDepartureService(command);

        mIsFollowing = true;
        updateFollowState();
        invalidateOptionsMenu();
    }

    private void startBoardedDepartureService(TripServiceCommand command) {
        Intent intent = new Intent(this, BoardedDepartureService.class)
                .setAction(command.getAction());
        if (command.getDeparture() != null) {
            intent.putExtra(BoardedDepartureService.DEPARTURE_EXTRA,
                    new DepartureParcel(command.getDeparture()));
        }
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            startForegroundService(intent);
        } else {
            startService(intent);
        }
    }

    private void requestNotificationPermissionIfNeeded() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU
                && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS)
                != android.content.pm.PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.POST_NOTIFICATIONS},
                    POST_NOTIFICATIONS_REQUEST_CODE);
        }
    }

    private void updateFollowState() {
        mFollowTripButton.setVisibility(mIsFollowing ? View.GONE : View.VISIBLE);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(mIsFollowing
                    ? R.string.trip_in_progress : R.string.train_schedule);
        }
    }

    private void renderTrip() {
        if (mDeparture == null || mDeparture.getStationPair() == null) {
            return;
        }
        mRoute.setText(mDeparture.getStationPair().getOrigin().getName() + " → "
                + mDeparture.getStationPair().getDestination().getName());
        mStatus.setText(getTripStatus());
        String estimatedArrival = DepartureTextFormatter.estimatedArrivalTime(
                this, mDeparture, false);
        mArrival.setText(estimatedArrival.isEmpty()
                ? getString(R.string.trip_final_arrival_unknown,
                mDeparture.getStationPair().getDestination().getName())
                : getString(R.string.trip_final_arrival,
                mDeparture.getStationPair().getDestination().getName(),
                estimatedArrival));

        mTimeline.removeAllViews();
        List<TripLeg> legs = mDeparture.getTripLegs();
        if (legs.isEmpty()) {
            addText(R.string.trip_details_unavailable, false);
            return;
        }
        int firstActiveLeg = getFirstActiveLegIndex(legs);
        for (int i = firstActiveLeg; i < legs.size(); i++) {
            TripLeg leg = legs.get(i);
            addLeg(leg, i == firstActiveLeg);
            if (i + 1 < legs.size() && i + 1 >= firstActiveLeg) {
                addConnection(leg, legs.get(i + 1));
            }
        }
    }

    private String getTripStatus() {
        if (mDeparture.isCanceled()) {
            return getString(R.string.trip_canceled);
        }
        long nowMillis = mTimeSource.nowMillis();
        if (!mDeparture.hasDeparted(nowMillis)) {
            return getString(R.string.trip_leaves_in,
                    DepartureTextFormatter.countdown(this, mDeparture, nowMillis));
        }

        String connectionStatus = getWaitingConnectionStatus();
        if (connectionStatus != null) {
            return connectionStatus;
        }

        TripStop nextStop = getNextStop();
        if (nextStop != null) {
            return getString(R.string.trip_next_stop, nextStop.getStation().getName(),
                    formatEta(nextStop.getArrivalTime()));
        }

        long arrival = mDeparture.getEstimatedArrivalTime();
        if (arrival > 0 && arrival <= nowMillis) {
            return getString(R.string.trip_arrived);
        }
        return getString(R.string.trip_current_train);
    }

    private int getFirstActiveLegIndex(List<TripLeg> legs) {
        long now = mTimeSource.nowMillis();
        for (int i = 0; i < legs.size(); i++) {
            if (!hasPassedAllStops(legs.get(i), now)) {
                return i;
            }
        }
        return legs.size();
    }

    private boolean hasPassedAllStops(TripLeg leg, long now) {
        if (!leg.getStops().isEmpty()) {
            for (TripStop stop : leg.getStops()) {
                if (stop.getArrivalTime() <= 0 || stop.getArrivalTime() > now) {
                    return false;
                }
            }
            return true;
        }
        return leg.getArrivalTime() > 0 && leg.getArrivalTime() <= now;
    }

    private TripStop getNextStop() {
        long now = mTimeSource.nowMillis();
        List<TripLeg> legs = mDeparture.getTripLegs();
        int firstActiveLeg = getFirstActiveLegIndex(legs);
        for (int i = firstActiveLeg; i < legs.size(); i++) {
            TripLeg leg = legs.get(i);
            for (TripStop stop : leg.getStops()) {
                if (stop.getStation() != leg.getOrigin()
                        && stop.getArrivalTime() > now) {
                    return stop;
                }
            }
        }
        return null;
    }

    private String getWaitingConnectionStatus() {
        long now = mTimeSource.nowMillis();
        List<TripLeg> legs = mDeparture.getTripLegs();
        for (int i = 0; i + 1 < legs.size(); i++) {
            TripLeg arrivingLeg = legs.get(i);
            TripLeg nextLeg = legs.get(i + 1);
            if (arrivingLeg.getArrivalTime() > 0
                    && arrivingLeg.getArrivalTime() <= now
                    && nextLeg.getDepartureTime() > now) {
                String lineName = nextLeg.getLine() == null ? "Train"
                        : nextLeg.getLine().getDisplayName();
                return getString(R.string.trip_transfer_now, lineName);
            }
            if (arrivingLeg.getArrivalTime() > 0
                    && arrivingLeg.getArrivalTime() <= now
                    && nextLeg.getDepartureTime() <= 0) {
                return getString(R.string.trip_no_departure_scheduled);
            }
        }
        return null;
    }

    private void addLeg(TripLeg leg, boolean isCurrentTrain) {
        TextView heading = addText(null, true);
        String lineName = leg.getLine() == null ? "Train"
                : leg.getLine().getDisplayName();
        String prefix = isCurrentTrain ? getString(R.string.trip_current_train)
                : getString(R.string.trip_connection_train);
        heading.setText(prefix + " · " + lineName);

        TextView route = addText(null, false);
        route.setText(leg.getOrigin().getName() + " → "
                + leg.getDestination().getName());
        route.setTextColor(getResources().getColor(R.color.text_secondary));

        if (leg.getStops().isEmpty()) {
            TextView arrival = addText(null, false);
            arrival.setText(leg.getDepartureTime() > 0
                    ? getString(R.string.trip_train_departure,
                    formatTime(leg.getDepartureTime()))
                    : getString(R.string.trip_no_departure_scheduled));
            return;
        }
        for (TripStop stop : leg.getStops()) {
            LinearLayout row = new LinearLayout(this);
            row.setOrientation(LinearLayout.HORIZONTAL);
            row.setPadding(dp(8), dp(7), dp(8), dp(7));

            TextView station = new TextView(this);
            station.setText(stop.getStation().getName());
            station.setTextColor(getResources().getColor(R.color.text_primary));
            station.setTextSize(16);
            row.addView(station, new LinearLayout.LayoutParams(0,
                    LinearLayout.LayoutParams.WRAP_CONTENT, 1));

            TextView time = new TextView(this);
            time.setGravity(android.view.Gravity.RIGHT);
            time.setTextColor(getResources().getColor(R.color.text_secondary));
            long displayedTime = stop.getStation() == leg.getOrigin()
                    ? stop.getDepartureTime() : stop.getArrivalTime();
            time.setText(formatTime(displayedTime) + "\n"
                    + formatEta(displayedTime));
            row.addView(time, new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT));
            mTimeline.addView(row);
        }
    }

    private void addConnection(TripLeg arrivingLeg, TripLeg nextLeg) {
        TextView connection = addText(null, true);
        connection.setText(getString(R.string.trip_transfer_at,
                arrivingLeg.getDestination().getName()));
        connection.setTextColor(getResources().getColor(R.color.brand_primary));

        TextView details = addText(null, false);
        long arrival = getArrivalAt(arrivingLeg, arrivingLeg.getDestination());
        long departure = getDepartureAt(nextLeg, nextLeg.getOrigin());
        if (departure <= 0) {
            details.setText(R.string.trip_no_departure_scheduled);
        } else {
            details.setText(getString(R.string.trip_connection_details,
                    formatTime(arrival), formatEta(arrival), formatTime(departure),
                    formatMargin(departure - arrival)));
        }
        long margin = departure - arrival;
        details.setTextColor(getResources().getColor(
                isConnectionWarning(arrivingLeg, margin)
                        ? R.color.connection_warning : R.color.text_secondary));
    }

    private boolean isConnectionWarning(TripLeg arrivingLeg, long margin) {
        int minimumSeconds = arrivingLeg.getMinimumTransferSecondsAfter();
        return margin < 0
                || (minimumSeconds > 0 && margin < minimumSeconds * 1000L);
    }

    private long getArrivalAt(TripLeg leg,
                              com.dougkeen.bart.model.Station station) {
        for (TripStop stop : leg.getStops()) {
            if (stop.getStation() == station) {
                return stop.getArrivalTime();
            }
        }
        return leg.getArrivalTime();
    }

    private long getDepartureAt(TripLeg leg,
                                com.dougkeen.bart.model.Station station) {
        for (TripStop stop : leg.getStops()) {
            if (stop.getStation() == station) {
                return stop.getDepartureTime();
            }
        }
        return leg.getDepartureTime();
    }

    private TextView addText(Integer resource, boolean heading) {
        TextView text = new TextView(this);
        text.setPadding(dp(8), heading ? dp(12) : dp(3), dp(8), dp(3));
        text.setTextSize(heading ? 18 : 14);
        text.setTextColor(getResources().getColor(
                heading ? R.color.text_primary : R.color.text_secondary));
        if (heading) {
            text.setTypeface(Typeface.DEFAULT, Typeface.BOLD);
        }
        if (resource != null) {
            text.setText(resource);
        }
        mTimeline.addView(text);
        return text;
    }

    private String formatTime(long time) {
        return time > 0 ? DateFormat.getTimeFormat(this)
                .format(new Date(time)) : "--";
    }

    private String formatEta(long time) {
        if (time <= 0) {
            return getString(R.string.trip_eta_unknown);
        }
        long seconds = (time - mTimeSource.nowMillis()) / 1000L;
        if (seconds <= 0) {
            return getString(R.string.trip_stop_passed);
        }
        long minutes = seconds / 60L;
        return getString(R.string.trip_eta_minutes_seconds, minutes,
                seconds % 60L);
    }

    private String formatMargin(long margin) {
        if (margin < 0) {
            return getString(R.string.trip_connection_missed);
        }
        long seconds = margin / 1000L;
        long minutes = seconds / 60L;
        seconds %= 60L;
        if (minutes == 0) {
            return getString(R.string.trip_margin_seconds, seconds);
        }
        return seconds == 0
                ? getString(R.string.trip_margin_minutes, minutes)
                : getString(R.string.trip_margin_minutes_seconds, minutes,
                seconds);
    }

    private int dp(int value) {
        return (int) (value * getResources().getDisplayMetrics().density + .5f);
    }

}

File: app/src/main/kotlin/com/dougkeen/bart/activities/TripInProgressActivity.kt
package com.dougkeen.bart.activities

import android.Manifest
import android.app.AlarmManager
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Build
import android.provider.Settings
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.viewModels
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.edit
import androidx.core.net.toUri
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.dougkeen.bart.BartRunnerApplication
import com.dougkeen.bart.R
import com.dougkeen.bart.model.Departure
import com.dougkeen.bart.model.Station
import com.dougkeen.bart.model.StationPair
import com.dougkeen.bart.presentation.DepartureTextFormatter
import com.dougkeen.bart.receivers.AlarmBroadcastReceiver
import com.dougkeen.bart.services.BoardedDepartureService
import com.dougkeen.bart.ui.BartRunnerTheme
import com.dougkeen.bart.ui.TripScreen
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class TripInProgressActivity : ComponentActivity() {
    companion object {
        private const val POST_NOTIFICATIONS_REQUEST_CODE = 1002
    }

    private val tripProgressViewModel: TripProgressViewModel by viewModels()
    private val tripActionsViewModel: TripActionsViewModel by viewModels()
    private var isFollowing = false
    private var routeDestination: Station? = null
    private var tripRoute: StationPair? = null
    private var tripFare by mutableStateOf<String?>(null)
    private var fareEligible = false
    private var fareLookupKey: String? = null
    private var pendingAlarmLeadTimeMinutes: Int? = null
    private var alarmVisible by mutableStateOf(false)

    override fun onResume() {
        super.onResume()
        tripActionsViewModel.refreshAlarmState()
        val pendingLeadTime = pendingAlarmLeadTimeMinutes
        if (pendingLeadTime != null && hasExactAlarmPermission()) {
            pendingAlarmLeadTimeMinutes = null
            enableAlarm(pendingLeadTime)
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        if (intent.getBooleanExtra(AlarmBroadcastReceiver.EXTRA_ALARM_TRIGGERED, false)) {
            alarmVisible = true
            showAlarmWindow()
            NotificationManagerCompat.from(this)
                .cancel(AlarmBroadcastReceiver.ALARM_NOTIFICATION_ID)
        }
        if (intent.getStringExtra(RouteArguments.DEPARTURE_IDENTITY) != null) {
            setIntent(intent)
            recreate()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val app = application as BartRunnerApplication
        val followed = tripActionsViewModel.getFollowedDeparture()
        var route = RouteArguments.readRoute(intent)
        var identity = RouteArguments.readDepartureIdentity(intent)
        val screenMode = RouteArguments.readScreenMode(intent)
        if (route == null && followed != null && (screenMode == null || screenMode == RouteArguments.MODE_FOLLOWED)) {
            identity = followed.identity
            route = followed.getStationPair()
        }
        if (route == null || identity == null) {
            finish()
            return
        }
        routeDestination = route.destination
        tripRoute = route
        fareEligible = route.destination != null
        tripFare = null
        resolveFare(app, route)
        alarmVisible = intent.getBooleanExtra(
            AlarmBroadcastReceiver.EXTRA_ALARM_TRIGGERED,
            false,
        )
        NotificationManagerCompat.from(this)
            .cancel(com.dougkeen.bart.receivers.AlarmBroadcastReceiver.ALARM_NOTIFICATION_ID)
        if (alarmVisible) {
            showAlarmWindow()
        }
        isFollowing = followed != null && identity == followed.identity
        tripProgressViewModel.setQuery(
            route,
            identity,
            app.timeSource,
            initialDeparture = if (isFollowing) followed else null,
        )

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                tripProgressViewModel.departureState.collect { updated ->
                    if (updated != null) {
                        if (isFollowing) tripActionsViewModel.updateFollowedTrip(updated)
                        resolveFare(app, updated.getStationPair())
                    }
                }
            }
        }

        setContent {
            val departure by tripProgressViewModel.departureState.collectAsStateWithLifecycle()
            val tripActionsState by tripActionsViewModel.uiState.collectAsStateWithLifecycle()
            BartRunnerTheme {
                TripScreen(
                    departure = departure,
                    route = tripRoute,
                    fare = tripFare,
                    isFollowingInitially = isFollowing,
                    alarmVisible = alarmVisible,
                    timeSource = app.timeSource,
                    alarmPending = tripActionsState.alarmPending,
                    alarmLeadTimeMinutes = tripActionsState.alarmLeadTimeMinutes,
                    onBack = { finish() },
                    onFollow = { followTrip(it) },
                    onSetAlarm = ::enableAlarm,
                    onCancelAlarm = {
                        tripActionsViewModel.cancelAlarm()
                        stopAlarmTrackingService()
                    },
                    onClear = {
                        tripActionsViewModel.clearTrip()
                        stopAlarmTrackingService()
                        finish()
                    },
                    onShare = ::shareArrival,
                    onSilenceAlarm = ::silenceAlarm,
                )
            }
        }
    }

    private fun silenceAlarm() {
        alarmVisible = false
        window.clearFlags(android.view.WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    private fun showAlarmWindow() {
        setShowWhenLocked(true)
        setTurnScreenOn(true)
        window.addFlags(android.view.WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    private fun enableAlarm(leadTimeMinutes: Int) {
        if (!hasExactAlarmPermission()) {
            pendingAlarmLeadTimeMinutes = leadTimeMinutes
            openExactAlarmSettings()
            return
        }
        if (!hasFullScreenIntentPermission()) {
            pendingAlarmLeadTimeMinutes = leadTimeMinutes
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                openFullScreenIntentSettings()
            }
            return
        }

        tripActionsViewModel.setAlarm(leadTimeMinutes)
        startAlarmTrackingService()
    }

    private fun hasExactAlarmPermission(): Boolean {
        return getSystemService(AlarmManager::class.java)?.canScheduleExactAlarms() == true
    }

    private fun hasFullScreenIntentPermission(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            return true
        }
        return getSystemService(android.app.NotificationManager::class.java)
            ?.canUseFullScreenIntent() == true
    }

    private fun openExactAlarmSettings() {
        val settingsIntent = Intent(
            Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM,
            "package:$packageName".toUri(),
        )
        try {
            startActivity(settingsIntent)
        } catch (_: android.content.ActivityNotFoundException) {
            startActivity(
                Intent(
                    Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                    "package:$packageName".toUri(),
                ),
            )
        }
    }

    @RequiresApi(Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
    private fun openFullScreenIntentSettings() {
        val settingsIntent = Intent(
            Settings.ACTION_MANAGE_APP_USE_FULL_SCREEN_INTENT,
            "package:$packageName".toUri(),
        )
        try {
            startActivity(settingsIntent)
        } catch (_: android.content.ActivityNotFoundException) {
            startActivity(
                Intent(
                    Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                    "package:$packageName".toUri(),
                ),
            )
        }
    }

    private fun followTrip(departure: Departure) {
        if (isFollowing) return
        tripActionsViewModel.followTrip(departure, tripRoute?.destination)
        requestNotificationPermissionIfNeeded()
        isFollowing = true
    }

    private fun resolveFare(app: BartRunnerApplication, route: StationPair?) {
        if (!fareEligible || tripFare != null) return
        val origin = route?.origin ?: return
        val destination = route.destination ?: return
        val lookupKey = "${origin.abbreviation}>${destination.abbreviation}"
        if (fareLookupKey == lookupKey) return
        fareLookupKey = lookupKey
        lifecycleScope.launch {
            val fare = withContext(Dispatchers.IO) {
                runCatching { app.gtfsStaticData.getFare(origin, destination) }.getOrNull()
            }
            if (fare != null && fareLookupKey == lookupKey) {
                tripFare = fare
            }
        }
    }

    private fun shareArrival(departure: Departure) {
        val destination = departure.getStationPair()?.destination
            ?: routeDestination
            ?: departure.trainDestination
            ?: return
        val message = getString(
            R.string.arrival_message,
            destination.getName(),
            DepartureTextFormatter.estimatedArrivalTime(this, departure, false),
        )
        startActivity(Intent.createChooser(Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_trip_subject))
            putExtra(Intent.EXTRA_TEXT, message)
        }, getString(R.string.share_arrival_time)))
    }

    private fun startAlarmTrackingService() {
        startForegroundService(
            Intent(this, BoardedDepartureService::class.java)
                .setAction(BoardedDepartureService.ACTION_START_ALARM_TRACKING),
        )
    }

    private fun stopAlarmTrackingService() {
        stopService(Intent(this, BoardedDepartureService::class.java))
    }

    private fun requestNotificationPermissionIfNeeded() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU &&
            ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED
        ) {
            requestPermissions(arrayOf(Manifest.permission.POST_NOTIFICATIONS), POST_NOTIFICATIONS_REQUEST_CODE)
        }
    }
}
